src/main/scala/com/wavesplatform/http/ApiMarshallers.scala

```diff
@@ -1,58 +1,66 @@
 package com.wavesplatform.http
 
-import scala.util.control.Exception.nonFatalCatch
-import scala.util.control.NoStackTrace
 import akka.http.scaladsl.marshalling.{Marshaller, PredefinedToEntityMarshallers, ToEntityMarshaller, ToResponseMarshaller}
 import akka.http.scaladsl.model.MediaTypes.{`application/json`, `text/plain`}
-import akka.http.scaladsl.model.StatusCode
+import akka.http.scaladsl.model.{MessageEntity, StatusCode}
 import akka.http.scaladsl.unmarshalling.{FromEntityUnmarshaller, PredefinedFromEntityUnmarshallers, Unmarshaller}
 import akka.util.ByteString
-import play.api.libs.json._
 import com.wavesplatform.api.http.ApiError
 import com.wavesplatform.transaction.{Transaction, ValidationError}
+import play.api.libs.json._
+
+import scala.concurrent.Future
+import scala.util.control.Exception.nonFatalCatch
+import scala.util.control.NoStackTrace
 
 case class PlayJsonException(cause: Option[Throwable] = None, errors: Seq[(JsPath, Seq[JsonValidationError])] = Seq.empty)
     extends IllegalArgumentException
     with NoStackTrace
 
 trait ApiMarshallers {
-  type TRM[A] = ToResponseMarshaller[A]
-
   import akka.http.scaladsl.marshalling.PredefinedToResponseMarshallers._
-  implicit val aem: TRM[ApiError] = fromStatusCodeAndValue[StatusCode, JsValue].compose { ae =>
-    ae.code -> ae.json
-  }
-  implicit val vem: TRM[ValidationError] = aem.compose(ve => ApiError.fromValidationError(ve))
 
-  implicit val tw: Writes[Transaction] = Writes(_.json())
+  implicit lazy val ApiErrorMarshaller: ToResponseMarshaller[ApiError] =
+    fromStatusCodeAndValue[StatusCode, JsValue].compose(ae => (ae.code, ae.json))
+
+  implicit lazy val ValidationErrorMarshaller: ToResponseMarshaller[ValidationError] =
+    ApiErrorMarshaller.compose(ve => ApiError.fromValidationError(ve))
+
+  implicit lazy val TransactionJsonWrites: Writes[Transaction] = Writes(_.json())
 
-  private val jsonStringUnmarshaller =
+  private[this] lazy val jsonStringUnmarshaller =
     Unmarshaller.byteStringUnmarshaller
       .forContentTypes(`application/json`)
       .mapWithCharset {
         case (ByteString.empty, _) => throw Unmarshaller.NoContentException
         case (data, charset)       => data.decodeString(charset.nioCharset.name)
       }
 
-  private lazy val jsonStringMarshaller =
+  private[this] lazy val jsonStringMarshaller =
     Marshaller.stringMarshaller(`application/json`)
 
-  implicit def playJsonUnmarshaller[A](implicit reads: Reads[A]): FromEntityUnmarshaller[A] =
-    jsonStringUnmarshaller map { data =>
-      val json = nonFatalCatch.withApply(t => throw PlayJsonException(cause = Some(t)))(Json.parse(data))
+  implicit def playJsonUnmarshaller[A](implicit reads: Reads[A]): FromEntityUnmarshaller[A] = {
+    jsonStringUnmarshaller.andThen(Unmarshaller(implicit ec =>
```

Comment:
This looks just like map, which transforms unmarshaller's output on the provided execution context